### PR TITLE
Provides NixOS modules and VM Tests for plutus services

### PR DESCRIPTION
**Pull request by gilligan**
**Summary**:
Extract NixOS modules and provide tests for all services

**Details**
This PR extracts proper NixOS modules from the morph deployment configuration:
- marlowe-playground.nix
- plutus-playground.nix
- pab.nix
- webghc.nix

All of the above services were previously hardwired into the morph deployment configuration making them hard to use, test and reason about.

**NixOS Tests**
Additionally this PR provides [NixOS Tests](https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests). Individual tests for asserting the correctness of the respective service and `all.nix` which starts several dedicated machines with services running like on production. The following assertions are being made:

- check systemd service states
- check if ports are open
- check if frontends are being served (by grepping for some html content)
- check if fontends can invoke ghc compilation using the webghc service (by going via `/api/contracts` or `/runghc`)

----

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -19,4 +19,4 @@ jobs:
...
-      - run: nix-build -A tests --arg supportedSystems '[ builtins.currentSystem ]' --restrict-eval -I . --allowed-uris 'https://github.com/NixOS/nixpkgs https://github.com/input-output-hk https://github.com/NixOS/nixpkgs-channels' --option trusted-public-keys "hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= iohk.cachix.org-1:DpRUyj7h7V830dp/i6Nti+NEO2/nhblbov/8MW7Rqoo= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=" --option substituters "https://hydra.iohk.io https://iohk.cachix.org https://cache.nixos.org/"
+      - run: nix-build -A tests.nixpkgsFmt -A tests.purty -A tests.shellcheck -A tests.stylishHaskell -A tests.terraform --arg supportedSystems '[ builtins.currentSystem ]' --restrict-eval -I . --allowed-uris 'https://github.com/NixOS/nixpkgs https://github.com/input-output-hk https://github.com/NixOS/nixpkgs-channels' --option trusted-public-keys "hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= iohk.cachix.org-1:DpRUyj7h7V830dp/i6Nti+NEO2/nhblbov/8MW7Rqoo= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=" --option substituters "https://hydra.iohk.io https://iohk.cachix.org https://cache.nixos.org/"
```
Comment (gilligan) [inline .github/workflows/test.yml:22]: the github action vms can't do virtualization so we'll have to skip that one here.

File: nix/modules/marlowe-playground.nix [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        ProtectKernelModules = true;
+        PrivateDevices = true;
+        SystemCallArchitectures = "native";
+
+      };
+
+      path = [ pkgs.z3 killallz3 ];
+      script = ''
+        if [ -f /var/lib/playgrounds/marlowe.env ]; then
+          echo "Loading environment config from '/var/lib/playgrounds/marlowe.env'"
```
Comment (michaelpj) [inline nix/modules/marlowe-playground.nix:79]: Should this instead be an option? Either the path, or even the file iteslf. Or does it need to be discovered at runtime?

Comment (michaelpj) [reply nix/modules/marlowe-playground.nix:79]: Also we don't do anything with this path? I guess the server looks for it?

File: nix/modules/pab.nix [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        getWallet = "${builtins.toString cfg.defaultWallet}";
+      };
+    };
+
+    metadataServerConfig = {
+      mdBaseUrl = "http://localhost:${builtins.toString cfg.metadataPort}";
+    };
+  };
+
+  pabYaml = pkgs.writeText "pab.yaml" (builtins.toJSON pabConfig);
```
Comment (michaelpj) [inline nix/modules/pab.nix:64]: neat

File: nix/modules/pab.nix [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        # Sane defaults for security
+        ProtectKernelTunables = true;
+        ProtectControlGroups = true;
+        ProtectKernelModules = true;
+        PrivateDevices = true;
+        SystemCallArchitectures = "native";
+      };
+      postStart = ''
+        mkdir -p /var/lib/pab
+        ${lib.concatMapStringsSep "\n" (p: "${cfg.pab-package}/bin/plutus-pab --config=${pabYaml} contracts install --path ${p}") cfg.contracts}
```
Comment (michaelpj) [inline nix/modules/pab.nix:202]: Worth a comment about what's going on here.

File: nix/tests/vm-tests/all.nix [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+          port = 4000;
+          playground-server-package = plutus-playground.server;
+          webghcURL = "http://webghc";
+        };
+
+        nginx = {
+          enable = true;
+          recommendedGzipSettings = true;
+          recommendedProxySettings = true;
+          recommendedOptimisation = true;
```
Comment (michaelpj) [inline nix/tests/vm-tests/all.nix:93]: probably not worth it for the tests

File: nix/tests/vm-tests/all.nix [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+          upstreams = {
+            plutus-playground.servers."127.0.0.1:4000" = { };
+            marlowe-playground.servers."127.0.0.1:4001" = { };
+            marlowe-dashboard.servers."192.168.1.1:8080" = { };
+          };
+          virtualHosts = {
+            "marlowe-dashboard" = {
```
Comment (michaelpj) [inline nix/tests/vm-tests/all.nix:101]: We can't include these snippets from the normal setups for the various apps, can we?

**Review by michaelpj — APPROVED** [review]
This all looks pristine. Lovely to have some actual VM tests! Mostly just questions about whether we can reuse config etc.

File: nix/modules/marlowe-playground.nix (hunk 2 above)
Comment (gilligan) [reply nix/modules/marlowe-playground.nix:79]: This path will contain environment variables that provide secrets. They are going to be uploaded by `morph` during deployment. The service does however also work without these secrets hence they are optional here. The service gets restarted by morph after they have been uploaded.

File: nix/tests/vm-tests/all.nix (hunk 5 above)
Comment (gilligan) [reply nix/tests/vm-tests/all.nix:93]: it's not relevant, but i suppose it doesn't hurt either - i wanted reuse as much as possible from the existing production config

File: nix/tests/vm-tests/all.nix (hunk 6 above)
Comment (gilligan) [reply nix/tests/vm-tests/all.nix:101]: eventually yes - not now though. These NixOS services need to first find their way into the actual configs for production.

**Pull request merged**